File: _complete_archive/_archive_2026-02-26_18-50-32/stacks/r/base/code/task-trace.tpl.R
task_trace_path <- function() {
  p <- Sys.getenv('TASK_TRACE_PATH')
  if (nchar(p) == 0) {
    return(file.path('artifacts', 'task-trace.jsonl'))
  }
  p
}

.task_trace_write <- function(ev) {
  p <- task_trace_path()
  dir.create(dirname(p), recursive = TRUE, showWarnings = FALSE)
  if (is.null(ev$ts)) {
    ev$ts <- as.numeric(Sys.time())
  }
  line <- NULL
  if (requireNamespace('jsonlite', quietly = TRUE)) {
    line <- jsonlite::toJSON(ev, auto_unbox = TRUE, null = 'null')
  } else {
    line <- paste0('{"task_id":', dQuote(ev$task_id), ',"type":', dQuote(ev$type), ',"ts":', ev$ts, '}')
  }
  cat(paste0(line, "\n"), file = p, append = TRUE)
}

task_trace_emit <- function(task_id, type, name = NULL, key = NULL, table = NULL, keys = NULL, value = NULL, fields = NULL) {
  ev <- list(task_id = task_id, type = type)
  if (!is.null(name)) ev$name <- name
  if (!is.null(key)) ev$key <- key
  if (!is.null(table)) ev$table <- table
  if (!is.null(keys)) ev$keys <- keys
  if (!is.null(value)) ev$value <- value
  if (!is.null(fields)) {
    for (k in names(fields)) {
      ev[[k]] <- fields[[k]]
    }
  }
  .task_trace_write(ev)
}

File: _complete_archive/_archive_2026-02-26_18-50-32/stacks/typescript/base/code/task-trace.tpl.ts
import fs from 'fs';
import path from 'path';

type Fields = Record<string, unknown>;

function tracePath(): string {
  return process.env.TASK_TRACE_PATH || path.join('artifacts', 'task-trace.jsonl');
}

function writeEvent(ev: Record<string, unknown>): void {
  const p = tracePath();
  fs.mkdirSync(path.dirname(p), { recursive: true });
  const event = { ...ev, ts: (ev as any).ts ?? Date.now() / 1000 };
  fs.appendFileSync(p, JSON.stringify(event) + '\n', { encoding: 'utf8' });
}

export function emit(params: {
  task_id: string;
  type: string;
  name?: string;
  key?: string;
  table?: string;
  keys?: string[];
  value?: unknown;
  fields?: Fields;
}): void {
  const { task_id, type, name, key, table, keys, value, fields } = params;
  const ev: Record<string, unknown> = { task_id, type };
  if (name !== undefined) ev.name = name;
  if (key !== undefined) ev.key = key;
  if (table !== undefined) ev.table = table;
  if (keys !== undefined) ev.keys = keys;
  if (value !== undefined) ev.value = value;
  if (fields) Object.assign(ev, fields);
  writeEvent(ev);
}

export function emitEvent(task_id: string, name: string, fields?: Fields): void {
  emit({ task_id, type: 'event.emit', name, fields });
}

export function dbWrite(task_id: string, table: string, fields?: Fields): void {
  emit({ task_id, type: 'db.write', table, fields });
}

export function taskReturn(task_id: string, keys: string[], fields?: Fields): void {
  emit({ task_id, type: 'task.return', keys, fields });
}

export function logEmit(task_id: string, key: string, fields?: Fields): void {
  emit({ task_id, type: 'log.emit', key, fields });
}

export function metricEmit(task_id: string, name: string, value?: unknown, fields?: Fields): void {
  emit({ task_id, type: 'metric.emit', name, value, fields });
}
